src/lane_keep_assist/scripts/lane_keep_assist.py

```diff
@@ -1,57 +1,104 @@
 #!/usr/bin/env python3
 
 import rospy
-from std_msgs.msg import Float64, String
+import lane_turn
+from std_msgs.msg import Float64, UInt8
 from geometry_msgs.msg import PoseStamped
 from sensor_msgs.msg import PointCloud, Image
-
+from lane_keep_assist.msg import LaneStatus, Lane
 
 
 class LaneKeepAssist:
 
-    """
-    Keeps them lanes
-    """
     def __init__(self):
-        self.pub = rospy.Publisher('steering',Float64 ,queue_size=10)
+        self.steering_pub_test = rospy.Publisher('lane_info', LaneStatus, queue_size=10)
+        self.gradient_state = 1
+        self.hls_state = 1
+        self.segmentation_state = 1
+        self.gradient_lanes = []
+        self.hls_lanes = []
+        self.segmentation_lane = []
+        self.gradient_percent_diff = 0
+        self.hls_percent_diff = 0
+        self.image = None
+        self.segmented_image = None
+        self.road_colour = 0
+        self.first_detection = True
 
     def listener(self):
         rospy.init_node('LaneKeepAssist', anonymous=True)
         rospy.Subscriber("airsim/image_raw", Image, self.handle_camera_data)
-        rospy.Subscriber("carData", String, self.handle_car_data)
-        rospy.Subscriber("airsimPose", PoseStamped, self.handle_gps_data)
-        rospy.Subscriber("lidar", PointCloud, self.handle_lidar_data)
-        rospy.Subscriber("curve_data", String, self.handle_curve_data)
-        rospy.spin()
-
-    #Handles the camera data
-    def handle_camera_data(self, data):
-        print("Got camera data")
-
-    #Handles the GPS data
-    def handle_gps_data(self, data):
-        print("Got gps data")
-
-    #Handles the car data
-    def handle_car_data(self, data):
-        print("Got car data")
-
-    #Handles the lidar data
-    def handle_lidar_data(self, data):
-        print("Got lidar data")
-
-    #Handles the curve data
-    def handle_curve_data(self, data):
-        print("Got curve data")
-
-    def publish_results(self):
-        rate = rospy.Rate(2) # 2hz
+        rospy.Subscriber("segmented_image", Image, self.handle_segmented_image)
+
+        rate = rospy.Rate(10)
         while not rospy.is_shutdown():
-            self.pub.publish(1)
+            if self.image is not None and self.segmented_image is not None:
+                self.process()
+
+                # Lane status message carries info on the lane boundaries, road boundaries and the percent diff
+                lane_msg = LaneStatus()
+                lane_msg.lane_gradient_status = self.gradient_state
+                lane_msg.lane_hls_status = self.hls_state
+                lane_msg.lane_segmentation_status = self.segmentation_state
+                lane_msg.gradient_diff = self.gradient_percent_diff
```

Comment:
Hmm, not completely sure atm. Going straight in some sections of neighborhood gives  pretty good result, but I need to scrape through the log files to give a good answer for this. I doubt you would want to listen to a result that is greater than 30 or 40 percent.

Reply:
update on this, as weird as it seems you might be able to get away with an error of 100%, which is a value of 1. This looks odd but say one slope is 0.1 and the other is 0.18. When actually drawing that line the difference in line magnitude, at least visually, is not that large. However, it results in a percent error of around 80%. I would start with a threshold value of 1. I made it so that total disagreement is now 100.